src/aten/Copy.cpp

```diff
@@ -239,11 +244,42 @@ void _copy_xpu(TensorIterator& iter, bool non_blocking) {
 
   auto q = getCurrentSYCLQueue();
   if (non_blocking) {
-    // WA: Move back to asynchronization copy when HostCachingAlloctor supported
-    auto e = q.memcpy(dst, src, nbytes);
-    e.wait();
-    // TODO: If host tensor is pinned, we need record event in host caching
-    // allocator.
+    if (copy_kind == _H2D_) {
+      if (at::detail::getXPUHooks().isPinnedPtr(src)) {
+        q.memcpy(dst, src, nbytes);
+        at::xpu::CachingHostAllocator_recordEvent(
+            const_cast<void*>(src),
+            iter.tensor(1).storage().data_ptr().get_context(),
+            at::xpu::getCurrentXPUStream());
+      } else {
+        // Using stage memory for async copy to avoid incorrect
+        // free of src host memory before async H2D copy. E.g. memory allocated
+        // by CPU tensor factory won't be cached in CPU allocator. When host
+        // memory is freed with CPU tensor dtor at the end of train main loop,
+        // but the corresponding H2D copy might not have been executed yet.
+        auto stage_mem_dptr = at::xpu::HostAlloc(nbytes);
+        void* stage_mem = stage_mem_dptr.get();
+        if (!stage_mem) {
+          throw std::runtime_error(
+              "Fail to allocate host memory from XPU HostAllocator");
+        }
+
+        std::memcpy(stage_mem, src, nbytes);
+        q.memcpy(dst, stage_mem, nbytes);
+        at::xpu::CachingHostAllocator_recordEvent(
+            const_cast<void*>(stage_mem),
+            stage_mem_dptr.get_context(),
+            at::xpu::getCurrentXPUStream());
+      }
+    } else {
+      q.memcpy(dst, src, nbytes);
+      if (at::detail::getXPUHooks().isPinnedPtr(dst)) {
+        at::xpu::CachingHostAllocator_recordEvent(
```

Comment:
@guangyey   when will we abstract and use the unified API for all device types?

Reply:
After the device allocator...